<a href="https://colab.research.google.com/github/BridgingAISocietySummerSchools/Data-Science-AI-Python-Course/blob/main/02_advanced_self_learning/01_python_basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📓 Notebook 1 — Python Basics for Data Science

> **Advanced & Self-Learning Track — Notebook 1 of 10** · **Module:** Python Fundamentals · **Estimated time:** 30–35 min · **Difficulty:** Beginner

> ℹ️ This track is the *self-paced deep-dive* of the course. If you are looking for the 90-minute live **Introduction Session**, it lives in `01_introduction/`.

> 🔁 **Coming from the live Introduction Session?** Sections 1–2 (notebooks, variables) will feel familiar — treat them as a quick warm-up. The new depth starts in **Section 3**: type-conversion pitfalls, truncation vs. rounding, floor division with negative numbers, and floating-point surprises like `0.1 + 0.2 != 0.3`. **Starting cold?** Also fine — everything is explained from scratch.

Welcome to the first notebook of the advanced track! In data science you spend a *lot* of time storing values, formatting numbers and combining them into small calculations — long before you ever touch a machine-learning library. This notebook gets you fluent with the absolute basics so the rest of the course feels natural.

## 🎯 Learning objectives

By the end of this notebook you will be able to:

1. Run code and write notes inside a Jupyter notebook.
2. Use **variables** to store and update values.
3. Identify Python's core **data types**: `int`, `float`, `str`, `bool`, `None`.
4. Apply **arithmetic** and **string** operations confidently.
5. Format output with **f-strings**, including decimals, percentages and width.
6. Recognise common beginner pitfalls — truncation vs. rounding, floating-point surprises, type errors — and how to debug them.
7. Apply these basics to a small **investment-portfolio calculation**.

## ✅ Prerequisites

None — this is the very first notebook of the track. If you can read English and use a web browser, you are ready.

## 🧭 How this notebook is structured

Each section has three parts: **(1) intuition** in plain words, **(2) a runnable code cell**, and **(3) a short reflection** that explains the output. Exercises with full solutions appear at the end of every major block.

## 1. What is a Jupyter notebook?

A Jupyter notebook is an *interactive* document. It mixes:

- **Markdown cells** (like this one) for writing notes, equations and explanations, and
- **Code cells** that you can execute by pressing **Shift + Enter** (or clicking ▶ in Colab).

Think of it as a lab journal: you can write down a question, try the experiment, and keep the result side-by-side. This is exactly how working data scientists explore data.

Let's run our first code cell. Click it and press **Shift + Enter**:

In [1]:
# Anything after a "#" is a comment — Python ignores it.
# Comments are how you leave notes for your future self (and your teammates).

print("Hello, Python! 🐍")
print("Today is a great day to start data science.")

Hello, Python! 🐍
Today is a great day to start data science.


You should see two lines printed below the cell.
**Tip:** the order in which you *run* cells is what matters in a notebook — not the order in which they *appear*. If something behaves oddly, choose **Runtime → Restart and run all** (Colab) to make sure everything is fresh.

## 2. Variables — naming pieces of data

A **variable** is a name we attach to a value so we can reuse it later. Think of variables as *labelled boxes*:

```
┌──────────┐    ┌───────┐    ┌──────────┐
│  "Alice" │    │   25  │    │   5.6    │
└────┬─────┘    └───┬───┘    └─────┬────┘
     │              │              │
   name            age           height
```

In Python you create a variable by **assigning** a value with `=`:

In [2]:
name = "Alice"      # a string of text
age = 25            # an integer (whole number)
height = 1.70       # a float (decimal number)
is_student = True   # a boolean (True/False)

# Display all four neatly using f-strings (more on these later)
print(f"Name      : {name}")
print(f"Age       : {age}")
print(f"Height (m): {height}")
print(f"Student?  : {is_student}")

Name      : Alice
Age       : 25
Height (m): 1.7
Student?  : True


### Naming rules (and good style)

Python is strict about what is a *valid* name and the community has strong opinions about what is a *good* name:

| Rule                                                   | Example                |
|--------------------------------------------------------|------------------------|
| Must start with a letter or `_`                        | `age`, `_temp`         |
| Can contain letters, digits, `_`                       | `score_1`, `user_id`   |
| Case sensitive                                         | `Age` ≠ `age`          |
| Cannot be a reserved word (`for`, `class`, `if`, …)    | ❌ `class = 5`         |
| Use `snake_case` for variables and functions (PEP 8)   | ✅ `customer_count`    |

> 💡 **Choose names that describe meaning, not type.**  Prefer `temperature_celsius` over `t` or `temp_float`.

## 3. Core data types

In data science you'll meet four basic types over and over again:

| Type     | Python name | Example          | Used for                                 |
|----------|-------------|------------------|------------------------------------------|
| Integer  | `int`       | `42`             | counts, indices, IDs                      |
| Float    | `float`     | `19.99`          | measurements, probabilities, prices       |
| String   | `str`       | `"hello"`        | text, category labels, column names       |
| Boolean  | `bool`      | `True`, `False`  | flags, filters, comparison results        |

There is also a special placeholder, `None`, meaning "no value yet" — you'll see it for missing data later in the course.

In [3]:
# Use type() to ask Python: "what kind of value is this?"
count        = 42
price        = 19.99
city         = "Berlin"
has_discount = True
nothing      = None

for value in (count, price, city, has_discount, nothing):
    print(f"{repr(value):>10}  ->  {type(value).__name__}")

        42  ->  int
     19.99  ->  float
  'Berlin'  ->  str
      True  ->  bool
      None  ->  NoneType


**Reading the output.** Each row shows the value (in `repr()` form so the quotes around strings are visible) and the Python type name. Notice how `42` and `42.0` are *different* types — they look similar but behave subtly differently when you divide them.

### Converting between types

You can explicitly convert a value with `int()`, `float()`, `str()`, `bool()`. This is called *casting*:

In [4]:
# From string to number — common when reading CSV files
age_text = "25"
age_num  = int(age_text)
print(f"'25' as text  ->  {age_text!r:>6}  (type {type(age_text).__name__})")
print(f"25  as int    ->  {age_num:>6}  (type {type(age_num).__name__})")

# The string must "look like" the target type:
# int("3.9") would raise a ValueError — go via float() first:
print()
print(f"float('3.9')      = {float('3.9')}")
print(f"int(float('3.9')) = {int(float('3.9'))}")

# From float to int — note the loss of precision
print()
print(f"int(3.9)   = {int(3.9)}    <- truncates toward zero, does NOT round!")
print(f"int(-3.9)  = {int(-3.9)}   <- toward zero for negatives too")
print(f"round(3.9) = {round(3.9)}   <- this is what you usually want")

# round() on exact halves: Python rounds to the nearest EVEN number
print()
print(f"round(0.5) = {round(0.5)},  round(1.5) = {round(1.5)},  round(2.5) = {round(2.5)}")

# bool() in Python: 0, 0.0, '', None, [] are 'falsy'; almost everything else is 'truthy'
print()
print(f"bool(0)       = {bool(0)}")
print(f"bool('')      = {bool('')}")
print(f"bool('False') = {bool('False')}  <- non-empty string, so truthy!")

'25' as text  ->    '25'  (type str)
25  as int    ->      25  (type int)

float('3.9')      = 3.9
int(float('3.9')) = 3

int(3.9)   = 3    <- truncates toward zero, does NOT round!
int(-3.9)  = -3   <- toward zero for negatives too
round(3.9) = 4   <- this is what you usually want

round(0.5) = 0,  round(1.5) = 2,  round(2.5) = 2

bool(0)       = False
bool('')      = False
bool('False') = True  <- non-empty string, so truthy!


> ⚠️ **Four conversion pitfalls worth memorising.**
> 1. `int(3.9)` returns `3`, not `4` — casting *truncates toward zero* (so `int(-3.9)` is `-3`). Use `round()` when you actually want rounding.
> 2. `round()` has its own surprise: exact halves go to the nearest **even** number — `round(2.5)` is `2`, not `3`. This is *banker's rounding*: over many numbers it avoids a systematic upward bias (imagine a bank that always rounded interest up in its own favour…).
> 3. `int("3.9")` raises a `ValueError` — the string must look like an integer. Convert via `float()` first.
> 4. `bool("False")` is `True` — any *non-empty* string is truthy, regardless of what it says. When a CSV file hands you `"False"` as text, compare it (`value == "True"`) instead of casting it.

## 4. Arithmetic operations

Python supports the operators you would expect, plus two that are particularly useful in data science:

| Operator | Meaning             | Example         | Result |
|----------|---------------------|-----------------|--------|
| `+`      | addition            | `7 + 3`         | `10`   |
| `-`      | subtraction         | `7 - 3`         | `4`    |
| `*`      | multiplication      | `7 * 3`         | `21`   |
| `/`      | division (float!)   | `7 / 3`         | `2.333…` |
| `//`     | floor division      | `7 // 3`        | `2`    |
| `%`      | modulo (remainder)  | `7 % 3`         | `1`    |
| `**`     | exponent (power)    | `2 ** 10`       | `1024` |

`%` and `//` come up surprisingly often — for example to test whether a number is even (`n % 2 == 0`).

In [5]:
a, b = 17, 5

print(f"a + b  = {a + b}")
print(f"a - b  = {a - b}")
print(f"a * b  = {a * b}")
print(f"a / b  = {a / b:.4f}    <-- regular division always returns a float")
print(f"a // b = {a // b}        <-- floor division: drops the remainder")
print(f"a % b  = {a % b}        <-- remainder of a / b")
print(f"a ** b = {a ** b}    <-- a to the power b")

a + b  = 22
a - b  = 12
a * b  = 85
a / b  = 3.4000    <-- regular division always returns a float
a // b = 3        <-- floor division: drops the remainder
a % b  = 2        <-- remainder of a / b
a ** b = 1419857    <-- a to the power b


### Operator precedence — just like in maths

`*` and `/` bind tighter than `+` and `-`. Use parentheses whenever there is any doubt — it costs nothing and prevents bugs.

In [6]:
# These two are NOT the same!
print("2 + 3 * 4   =", 2 + 3 * 4)     # 14
print("(2 + 3) * 4 =", (2 + 3) * 4)   # 20

# In data work, parentheses make the *intent* explicit:
revenue = 1000
tax_rate = 0.19
net = revenue / (1 + tax_rate)
print(f"Net revenue: €{net:,.2f}")

2 + 3 * 4   = 14
(2 + 3) * 4 = 20
Net revenue: €840.34


## 5. Strings — working with text

Strings are how we hold text: names, labels, file paths, category values. Data is rarely "just numbers": a column of cities, a customer name, a status flag — they all show up as strings.

In [7]:
first_name = "Ada"
last_name  = "Lovelace"

# 1. Concatenation with +
full_name = first_name + " " + last_name
print(f"Full name : {full_name}")

# 2. Length
print(f"Letters    : {len(full_name)}")

# 3. Case methods (return NEW strings — they don't modify the original)
print(f"Upper      : {full_name.upper()}")
print(f"Lower      : {full_name.lower()}")
print(f"Title case : {full_name.title()}")

# 4. Stripping whitespace — VERY common when cleaning CSV data
messy = "   data science   "
print(f"Stripped   : {messy.strip()!r}")

# 5. Splitting and joining — also common in data cleaning
csv_row = "Ada,Lovelace,1815"
parts = csv_row.split(",")
print(f"Split      : {parts}")

Full name : Ada Lovelace
Letters    : 12
Upper      : ADA LOVELACE
Lower      : ada lovelace
Title case : Ada Lovelace
Stripped   : 'data science'
Split      : ['Ada', 'Lovelace', '1815']


> 💡 String methods return a *new* string. Strings in Python are **immutable** — `name.upper()` does not change `name`; it gives you back a new value that you must assign if you want to keep it.

## 6. F-strings — formatting output

F-strings (Python 3.6+) are the modern way to build strings that include variable values. They start with `f` and use `{ ... }` to embed expressions.

The general pattern is:

```
f"... {expression:format_spec} ..."
```

The `:format_spec` part is optional but very useful for clean output.

In [8]:
student = "Alice"
score   = 87.6543
pi      = 3.14159265

# Basic embedding
print(f"Hello, {student}!")

# Two-decimal places
print(f"Score : {score:.2f}")

# As a percentage (multiplies by 100 and adds %)
print(f"Pass rate : {0.873:.1%}")

# Width and alignment
print(f"|{'left':<10}|{'center':^10}|{'right':>10}|")
print(f"|{student:<10}|{score:^10.1f}|{pi:>10.4f}|")

# Thousands separator (great for money / large counts)
revenue = 1_234_567.89
print(f"Revenue : €{revenue:,.2f}")

# Debug form (Python 3.8+): show variable name + value
print(f"{score=:.2f}")

Hello, Alice!
Score : 87.65
Pass rate : 87.3%
|left      |  center  |     right|
|Alice     |   87.7   |    3.1416|
Revenue : €1,234,567.89
score=87.65


**Cheat-sheet of the most useful format specs**

| Spec       | Meaning                              | Example output |
|------------|--------------------------------------|----------------|
| `:.2f`     | fixed-point, 2 decimals              | `3.14`         |
| `:.1%`     | percent, 1 decimal                   | `12.3%`        |
| `:,`       | thousands separator                  | `1,234,567`    |
| `:>10`     | right-align in width 10              | `        42`   |
| `:<10`     | left-align in width 10               | `42        `   |
| `:^10`     | centre-align in width 10             | `    42    `   |
| `:.3e`     | scientific notation, 3 decimals      | `1.235e+03`    |

## 7. Updating variables

Variables can be reassigned. A very common shortcut is the *augmented assignment* operator `+=` (and friends):

In [9]:
balance = 1000
print(f"Start          : €{balance:,.2f}")

# Long form
balance = balance + 150
print(f"After +150     : €{balance:,.2f}")

# Augmented form — same thing, less typing
balance += 200   #  equivalent to: balance = balance + 200
print(f"After +200 more: €{balance:,.2f}")

# Works for the other operators too
counter = 1
counter *= 2   # counter = counter * 2
counter *= 2
counter *= 2
print(f"Counter doubled 3 times: {counter}")

Start          : €1,000.00
After +150     : €1,150.00
After +200 more: €1,350.00
Counter doubled 3 times: 8


## 8. A first applied example — investment portfolio

Let's combine everything: variables, arithmetic, strings and f-strings to compute the value and one-year dividend yield of a small portfolio. This is the *kind* of small calculation a financial analyst might do every day.

In [10]:
# === A tiny portfolio of three stocks ===
# (price per share, shares held, annual dividend per share)
stock_a_price, stock_a_shares, stock_a_dividend = 150.75, 50, 2.30
stock_b_price, stock_b_shares, stock_b_dividend = 89.20, 75, 1.45
stock_c_price, stock_c_shares, stock_c_dividend = 234.50, 25, 4.10

# Position value = price * shares
value_a = stock_a_price * stock_a_shares
value_b = stock_b_price * stock_b_shares
value_c = stock_c_price * stock_c_shares
total_value = value_a + value_b + value_c

# Dividend income = dividend per share * shares
div_a = stock_a_dividend * stock_a_shares
div_b = stock_b_dividend * stock_b_shares
div_c = stock_c_dividend * stock_c_shares
total_div = div_a + div_b + div_c

# Yield = annual dividends / total value
yield_pct = total_div / total_value

# === Pretty report ===
print("=" * 50)
print("📊 PORTFOLIO REPORT".center(50))
print("=" * 50)
print(f"{'Position':<10}{'Value':>15}{'Dividend':>15}")
print("-" * 50)
print(f"{'Stock A':<10}{value_a:>15,.2f}{div_a:>15,.2f}")
print(f"{'Stock B':<10}{value_b:>15,.2f}{div_b:>15,.2f}")
print(f"{'Stock C':<10}{value_c:>15,.2f}{div_c:>15,.2f}")
print("-" * 50)
print(f"{'TOTAL':<10}{total_value:>15,.2f}{total_div:>15,.2f}")
print(f"\nDividend yield: {yield_pct:.2%}")

                📊 PORTFOLIO REPORT                
Position            Value       Dividend
--------------------------------------------------
Stock A          7,537.50         115.00
Stock B          6,690.00         108.75
Stock C          5,862.50         102.50
--------------------------------------------------
TOTAL           20,090.00         326.25

Dividend yield: 1.62%


**What just happened?**

1. We stored 9 raw numbers as variables with clear names.
2. We *derived* new variables (`value_a`, `div_a`, …) using arithmetic — no copy-paste of numbers.
3. We assembled a nicely aligned report using f-string width and decimal specs.

Notice how every number on the report comes from a single calculation. If a stock price changes you only need to edit it in one place — that's the power of variables.

## 9. Common beginner pitfalls

Spotting these will save you hours of debugging later.

1. **Type vs. text.** `"25"` is a string, `25` is an integer. `"25" + 5` raises `TypeError`. Convert with `int("25")` first.
2. **Integer vs. float division.** `5 / 2` is `2.5`. `5 // 2` is `2`. With negatives the two rounding directions disagree: `//` *floors* (`-5 // 2` is `-3`) while `int()` *truncates* (`int(-5 / 2)` is `-2`). Choose deliberately.
3. **Strings are immutable.** `name.upper()` returns a new string. Always assign it back if you want to keep the change: `name = name.upper()`.
4. **Re-using variable names.** Assigning `list = [...]` overwrites the built-in `list` function. Avoid names that shadow built-ins (`list`, `dict`, `str`, `type`, `sum`, `id`, `input`).
5. **Comparing floats with `==`.** Because of binary floating-point, `0.1 + 0.2 == 0.3` is `False`. Use `abs(a - b) < 1e-9` for "close enough" comparisons.

Run the next cell to *see* a couple of these in action:

In [11]:
# Pitfall 2: division
print("5 / 2  =", 5 / 2)      # 2.5  (float)
print("5 // 2 =", 5 // 2)     # 2    (int — the remainder is dropped)

# ...and with NEGATIVE numbers, // and int() disagree:
print("-5 // 2     =", -5 // 2)       # -3  (floors: toward minus infinity)
print("int(-5 / 2) =", int(-5 / 2))   # -2  (truncates: toward zero)

# Pitfall 5: floating-point surprises
a, b = 0.1 + 0.2, 0.3
print(f"\n0.1 + 0.2 = {a!r}")
print(f"is it == 0.3?  {a == b}")
print(f"close enough?  {abs(a - b) < 1e-9}")

# Pitfall 1: TypeError on text + number
try:
    "25" + 5
except TypeError as e:
    print(f"\nTypeError caught: {e}")

5 / 2  = 2.5
5 // 2 = 2
-5 // 2     = -3
int(-5 / 2) = -2

0.1 + 0.2 = 0.30000000000000004
is it == 0.3?  False
close enough?  True

TypeError caught: can only concatenate str (not "int") to str


**Reading the output — why does `0.1 + 0.2` come out as `0.30000000000000004`?**

Computers store numbers in *binary*, and `0.1` is a repeating fraction in binary — just like `1/3 = 0.3333…` never terminates in decimal. The computer has to cut the digits off somewhere, and that tiny leftover error becomes visible when such numbers are added. This is not a Python bug; every mainstream language behaves the same way. Two practical consequences:

- never compare floats with `==` — use a tolerance, as in `abs(a - b) < 1e-9`;
- financial software often sidesteps the problem by counting in whole cents (integers) or using Python's `decimal` module.

The negative-division lines deserve a second look too: `-5 // 2` gives `-3` (floor division always rounds *down*, toward minus infinity) while `int(-5 / 2)` gives `-2` (casting always cuts *toward zero*). For positive numbers the two happen to agree — which is exactly why this bug hides until negative data shows up.

## 🧪 Practice exercises

Try each exercise yourself first. Solutions are right below — but learning happens *while struggling*, not while reading the answer.

### Exercise 1.1 — Personal data card

Create variables for your **name**, **age in years**, and **height in metres**. Print them in a sentence:
> *"Hi, I'm Alice, I'm 25 years old and 1.70 m tall."*

In [12]:
# Your code here  👇
my_name   = "Alice"
my_age    = 25
my_height = 1.70

# ...


<details>
<summary>💡 Click to reveal the solution</summary>

```python
my_name   = "Alice"
my_age    = 25
my_height = 1.70

print(f"Hi, I'm {my_name}, I'm {my_age} years old and {my_height:.2f} m tall.")
```

**Reasoning.** We use an f-string so each variable can be embedded directly. The `:.2f` spec on `my_height` keeps exactly two decimals (so `1.7` shows as `1.70`).
</details>

### Exercise 1.2 — Tip calculator

A restaurant bill is **€84.50** and you want to tip **18 %**. Compute and print:

1. the tip in euros (2 decimals),
2. the total to pay (2 decimals),
3. the per-person amount if the bill is split between 3 people.

In [13]:
# Your code here  👇
bill = 84.50
tip_rate = 0.18

# ...


<details>
<summary>💡 Click to reveal the solution</summary>

```python
bill      = 84.50
tip_rate  = 0.18
people    = 3

tip       = bill * tip_rate
total     = bill + tip
per_head  = total / people

print(f"Bill        : €{bill:>8.2f}")
print(f"Tip ({tip_rate:.0%})   : €{tip:>8.2f}")
print(f"Total       : €{total:>8.2f}")
print(f"Per person  : €{per_head:>8.2f}")
```

**Reasoning.** We store the inputs in clearly named variables, derive everything else, and format the column of euro amounts with width `8` so they line up.
</details>

### Exercise 1.3 — Average and pass / fail

You have three test scores. Compute the **average** and print whether it is a passing grade (≥ 70). For now, write the check as a simple comparison and `print` — we'll learn `if/else` in the next notebook.

In [14]:
# Your code here  👇
score1, score2, score3 = 78, 85, 92

# ...


<details>
<summary>💡 Click to reveal the solution</summary>

```python
score1, score2, score3 = 78, 85, 92

average = (score1 + score2 + score3) / 3
passed  = average >= 70

print(f"Average score : {average:.1f}")
print(f"Passed?       : {passed}")
```

**Reasoning.** The comparison `average >= 70` evaluates to a Boolean (`True`/`False`), which is itself a useful value. In notebook 2 we'll wrap it in an `if` statement to print a friendlier message.
</details>

### Exercise 1.4 — Predict the output 🔮

For each expression below, **write down your prediction first** (in the comments of the next cell), then reveal the answers and compare. A mismatch is gold: it points exactly at the concept to re-read. This *active recall* is the most effective way to learn without an instructor.

1. `int(7.99)`
2. `round(7.5)`
3. `round(8.5)`
4. `-7 // 2`
5. `0.1 + 0.2 == 0.3`
6. `bool("False")`

In [15]:
# Write your prediction after each arrow BEFORE checking 👇
# 1. int(7.99)         -> ?
# 2. round(7.5)        -> ?
# 3. round(8.5)        -> ?
# 4. -7 // 2           -> ?
# 5. 0.1 + 0.2 == 0.3  -> ?
# 6. bool("False")     -> ?

# Then uncomment this line to check all six at once:
# print(int(7.99), round(7.5), round(8.5), -7 // 2, 0.1 + 0.2 == 0.3, bool("False"))

<details>
<summary>💡 Click to reveal the solution</summary>

| Expression | Result | Why |
|------------|--------|-----|
| `int(7.99)` | `7` | casting truncates toward zero — it never rounds |
| `round(7.5)` | `8` | exact halves go to the nearest **even** number — here `8` |
| `round(8.5)` | `8` | same banker's-rounding rule — so `round(7.5)` and `round(8.5)` give the *same* answer! |
| `-7 // 2` | `-4` | floor division rounds *down* (toward minus infinity), not toward zero |
| `0.1 + 0.2 == 0.3` | `False` | binary floating-point: the sum is actually `0.30000000000000004` |
| `bool("False")` | `True` | any non-empty string is truthy — what the text *says* is irrelevant |

6/6 on the first try means you have genuinely absorbed the subtle parts of this notebook. Most people miss `round(8.5)` or `-7 // 2` — if you did, re-read Sections 3 and 9.
</details>

### Exercise 1.5 — Debug me 🐞

The commented-out code below contains **three bugs**. Read it carefully and predict what will happen. Then uncomment the lines, run the cell, read the error message, and fix the bugs so the cell prints a sensible message.

In [16]:
# Buggy — uncomment the lines below, run the cell, read the error, then fix the three bugs:
# revenue = "1200"
# costs = 800
# profit = revenue - costs
# margin = profit / revenue
# print("Profit: " + profit + ", margin: " + margin)

<details>
<summary>💡 Click to reveal the solution</summary>

The three bugs are:

1. `revenue = "1200"` — it's a string, not a number, so the arithmetic fails. Convert to `int` (or just write `1200`).
2. `"Profit: " + profit + …` — `profit` is a number, you cannot concatenate a string and a number. Use an f-string instead.
3. The same problem with `margin`, plus we should format it as a percentage.

```python
revenue = 1200          # 1️⃣ number, not string
costs   = 800
profit  = revenue - costs
margin  = profit / revenue
print(f"Profit: €{profit:,.2f}, margin: {margin:.1%}")   # 2️⃣, 3️⃣
```
</details>

## 🎁 Bonus mini-project — Compound interest

Compute the value of an investment after **n years** with compound interest:

$$V_n = V_0 \cdot (1 + r)^n$$

Use these inputs: an initial deposit of **€2 000**, an annual interest rate of **4.5 %**, over **10 years**. Print the final value, the total interest earned, and the growth factor. (A year-by-year *table* needs a loop — we'll build exactly that at the start of Notebook 2.)

In [17]:
# Your code here  👇
initial = 2000
rate = 0.045
years = 10

# Hint: value_after_n_years = initial * (1 + rate) ** years


<details>
<summary>💡 Click to reveal the solution</summary>

```python
initial = 2000
rate    = 0.045
years   = 10

final = initial * (1 + rate) ** years
gain  = final - initial

print(f"After {years} years at {rate:.1%} annual interest:")
print(f"   Final value  : €{final:,.2f}")
print(f"   Interest won : €{gain:,.2f}")
print(f"   Growth factor: ×{final/initial:.2f}")
```

We will revisit this example with a loop in Notebook 2 to print a year-by-year table.
</details>

## 🧠 Key takeaways

1. **Variables** are labelled containers — pick descriptive `snake_case` names.
2. The four core types you'll meet daily are `int`, `float`, `str`, `bool`. `None` represents "no value".
3. **Convert** between types with `int()`, `float()`, `str()`, `bool()`; `int()` truncates toward zero, and `round()` sends exact halves to the nearest *even* number (banker's rounding).
4. **Arithmetic** in Python supports `+ - * / // % **`. `/` always returns a float; `//` floors — it rounds *down* even for negatives (`-5 // 2` is `-3`).
5. **F-strings** are the standard way to format output. Master `.2f`, `,`, `%`, alignment specs — you will use them constantly.
6. Strings are **immutable** — methods return new strings. Always assign the result back if you want to keep it.
7. Watch out for **floating-point comparison** issues: prefer `abs(a - b) < eps` over `a == b` for floats.

## ✅ Self-assessment

Tick each box when you can do it without looking:

- [ ] Create variables of all four basic types
- [ ] Explain the difference between `5 / 2` and `5 // 2`
- [ ] Predict `round(2.5)` and `-7 // 2` — and explain *why*
- [ ] Convert a string to an integer and back
- [ ] Format a number as percentage and as a euro amount with thousands separator
- [ ] Recognise and fix at least 3 common type errors

## 🚀 Next step

Continue with **Notebook 2 — Control Structures** (`02_control_structures.ipynb`, in this folder) to learn how to make decisions (`if/else`) and repeat work (`for`, `while`). That's where Python really starts to feel like programming.